### PR TITLE
feat(sdk): Remove the need to “restart” `SlidingSync`

### PR DESCRIPTION
**Pull request by Hywan**
A large surface of bugs in `SlidingSync` is caused by the design of `SlidingSync` in general: when a list's configuration is modified, or when a new is added, when a subscription is made etc., `SlidingSync::stream` must be stopped, and started again. This design is called “restarting” `SlidingSync`.

First off, this is not needed. We can make `SlidingSync` works while it's syncing (while `stream` is running).

Second, it would improve safety around `SlidingSync` if we don't restart it. To be able to restart it, some states were stored in various types; `SlidingSyncList` was required to be clonable; and so on. Not only it will improve safety (like, a list cannot be cloned and used by 2 different `SlidingSync` instances), but it will improve performance.

## Progress

- [x] Make `SlidingSyncList` no longer clonable
- [x] `SlidingSync::add_list` and `SlidingSyncBuilder::add_list` takes a `SlidingSyncListBuilder` instead of a `SlidingSyncList`, so that it's easier to add private configurations to the lists
- [x] `SlidingSync` and `SlidingSyncList` are sharing a channel (sender + receiver)
- [x] `SlidingSync` can handle messages from multiple `SlidingSyncList` (with `tokio::select!`) and act accordingly
- [x] `SlidingSyncInternalMessage` has a variant for stopping the sync loop, do some stuff, and start it again
- [x] Use a “once built” callback on `SlidingSyncListBuilder` to configure a list once it's built, it's useful for FFI/Swift/Kotlin to set up observers

### CODE REVIEW

File: crates/matrix-sdk/src/sliding_sync/list/builder.rs [hunk 1]
```diff
@@ -17,7 +17,7 @@ use crate::Result;
 pub const FULL_SYNC_LIST_NAME: &str = "full-sync";
 
 /// Builder for [`SlidingSyncList`].
-#[derive(Clone, Debug)]
+#[derive(Debug, Clone)]
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync/list/builder.rs:20]:
```suggestion
#[derive(Clone, Debug)]
```

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 2]
```diff
@@ -118,6 +121,12 @@ pub struct Client {
     pub(crate) sliding_sync_reset_broadcast_tx: Arc<SharedObservable<()>>,
 }
 
+impl fmt::Debug for Client {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter.debug_struct("Client").field("client", &self.client).finish_non_exhaustive()
+    }
+}
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/client.rs:128]: What is this impl for?

File: crates/matrix-sdk/src/sliding_sync/builder.rs [hunk 3]
```diff
@@ -14,33 +14,28 @@ use ruma::{
...
+    SlidingSyncListBuilder, SlidingSyncPositionMarkers, SlidingSyncRoom,
 };
 use crate::{Client, Result};
 
 /// Configuration for a Sliding Sync instance.
 ///
 /// Get a new builder with methods like [`crate::Client::sliding_sync`], or
 /// [`crate::SlidingSync::builder`].
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/builder.rs:30]: Thanks!

File: crates/matrix-sdk/src/sliding_sync/list/builder.rs (hunk 1 above)
Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/list/builder.rs:20]: Thanks!

File: bindings/matrix-sdk-ffi/src/client.rs (hunk 2 above)
Comment (Hywan) [reply bindings/matrix-sdk-ffi/src/client.rs:128]: It was required at some point, it's no longer needed. I'm removing them.

File: bindings/matrix-sdk-ffi/src/sliding_sync.rs [hunk 4]
```diff
@@ -802,7 +788,7 @@ impl SlidingSync {
     }
 }
 
-#[derive(Clone, uniffi::Object)]
+#[derive(Debug, Clone, uniffi::Object)]
```
[the remark this answers is not in the document]
Comment (Hywan) [reply bindings/matrix-sdk-ffi/src/sliding_sync.rs:785]: Yup, old code. Cleaned up, thanks!

File: crates/matrix-sdk/src/sliding_sync/builder.rs [hunk 5]
```diff
@@ -71,10 +72,10 @@ impl SlidingSyncBuilder {
     /// Add the given list to the lists.
     ///
     /// Replace any list with the name.
-    pub fn add_list(mut self, list: SlidingSyncList) -> Self {
-        self.lists.insert(list.name().to_owned(), list);
+    pub fn add_list(mut self, list_builder: SlidingSyncListBuilder) -> Result<Self> {
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/builder.rs]: True!

**Comment by Hywan** [comment]
https://github.com/matrix-org/matrix-rust-sdk/pull/1881 will fix the Clippy error.
And the rest is a NodeJS error, will fix with another PR.

**Review by stefanceriu — APPROVED** [review]
Spent some time with Ivan today getting this to work properly and I'm happy to report that it totally does 🚀

**Pull request merged**